backend/src/routes/getHistoricalBalance.ts

```diff
@@ -0,0 +1,46 @@
+import { sql } from 'drizzle-orm';
+import type { FastifyInstance } from 'fastify';
+import { usdcBalance } from '../db/schema';
+
+const addressRegex = /^0x0[0-9a-fA-F]{63}$/;
+
+export function getHistoricalBalanceRoute(fastify: FastifyInstance) {
+  fastify.get('/get_historical_balance', async (request, reply) => {
+    const { address } = request.query as { address?: string };
+
+    if (!address) {
+      return reply.status(400).send({ error: 'Address is required' });
+    }
+
+    // Validate address format
+    if (!addressRegex.test(address)) {
+      return reply.status(400).send({ error: 'Invalid address format' });
+    }
+
+    try {
+      const historicalBalances = await fastify.db.execute(
+        sql`
+          SELECT balance, DATE(block_timestamp) AS date
+          FROM ${usdcBalance}
+          WHERE (address, block_timestamp) IN (
+            SELECT address, MAX(block_timestamp) AS max_timestamp
+            FROM ${usdcBalance}
+            WHERE address = ${address}
+            GROUP BY address, DATE(block_timestamp) 
+          )
+        `,
+      );
```

Comment:
don't add the query as raw sql use the ts functions

Reply:
also we want the last balance per day. If I had 100 usdc, i send 40 at 10 am then i receive 20 at 11pm my balance for that day would be 80 (lmk if it's not clear)

Reply:
and we want that for the last 30 days

Reply:
Okay. I assumed each entry in the database has a timestamp and corresponding balance. So I grouped the rows by address and date then extracted the entry with the latest timestamp per day assuming the entry already has the latest balance for that day.

Reply:
ah then this is correct, just rewrite this using the typescript functions plz

Reply:
Alright